src/lib.rs

```diff
@@ -96,20 +97,146 @@ pub struct SelectionOutput {
 pub fn select_coin_bnb(
     inputs: &[OutputGroup],
     options: CoinSelectionOpt,
+    rng: &mut ThreadRng,
 ) -> Result<SelectionOutput, SelectionError> {
-    unimplemented!()
+    let mut selected_inputs: Vec<usize> = vec![];
+    let bnb_tries = 1000000;
+
+    let mut sorted_inputs: Vec<(usize, OutputGroup)> = inputs
+        .iter()
+        .enumerate()
+        .map(|(index, input)| (index, input.clone()))
+        .collect();
+    sorted_inputs.sort_by_key(|(_, input)| std::cmp::Reverse(input.value));
+
+    let bnb_selected_coin = bnb(
+        &sorted_inputs,
+        &mut selected_inputs,
+        0,
+        0,
+        bnb_tries,
+        &options,
+        rng,
+    );
+    match bnb_selected_coin {
+        Some(selected_coin) => {
+            let accumulated_value: u64 = selected_coin
+                .iter()
+                .fold(0, |acc, &i| acc + inputs[i].value);
+            let accumulated_weight: u32 = selected_coin
+                .iter()
+                .fold(0, |acc, &i| acc + inputs[i].weight);
+            let estimated_fee = 0;
+            let waste = calculate_waste(
+                inputs,
+                &selected_inputs,
+                &options,
+                accumulated_value,
+                accumulated_weight,
+                estimated_fee,
+            );
+            let selection_output = SelectionOutput {
+                selected_inputs: selected_coin,
+                waste: WasteMetric(waste),
+            };
+            Ok(selection_output)
+        }
+        None => select_coin_srd(inputs, options, &mut rand::thread_rng()),
+    }
 }
 
 /// Return empty vec if no solutions are found
+// changing the selected_inputs : &[usize] -> &mut Vec<usize>
 fn bnb(
     inputs_in_desc_value: &[(usize, OutputGroup)],
-    selected_inputs: &[usize],
-    effective_value: u64,
+    selected_inputs: &mut Vec<usize>,
+    acc_eff_value: u64,
     depth: usize,
     bnp_tries: u32,
     options: &CoinSelectionOpt,
-) -> Vec<usize> {
-    unimplemented!()
+    rng: &mut ThreadRng,
+) -> Option<Vec<usize>> {
+    let target_for_match = options.target_value
+        + calculate_fee(options.base_weight, options.target_feerate)
+        + options.cost_per_output;
+    let match_range = options.cost_per_input + options.cost_per_output;
+    if acc_eff_value > target_for_match + match_range {
+        return None;
+    } else if acc_eff_value >= target_for_match {
+        return Some(selected_inputs.to_vec());
+    } else if bnp_tries <= 0 || depth >= inputs_in_desc_value.len() {
+        return None;
+    } else {
```

Comment:
A series of `if clauses` without the usage of` if-else ` clause would be more readable and also the `else` clause is redundant.